### DISCUSSION TITLE
[Idea] Improve category listing products loading by avoid duplicates products queries. (Prestashop 8 and below)

### DISCUSSION THREAD

**Discussion by MaximeOh** [PrestaShop/PrestaShop#33060] [Ideas]
### The problem

When you load products listing for a category by the function _**Category::getProducts()**_, a global request is made to get products information.

After, for each products, a lot of independent queries are made to "re-hydrate" the product by the function _**Product::getProductProperties()**_.

I identify a lot of queries could be avoid because the global category listing queries contains already the product data.
This problem is really noticeable when you have categories with lots of products.

### The solution

 it's possible to use it with a "pre-caching" step and avoid a lot of SQL queries.

Product's functions used some cache techniques, sometimes with internal static variables or with the Cache class.

Just before call the function Product::getProductProperties(), we can pre-fill the cache with the data present in the global query.

#### Cover

The cover is loaded by the static function **_Product::getCover()_**, if _"cover_image_id"_ is unset.
But this object is already loaded by the global query (_"id_image"_ and localized _"legend"_).

####  Tax Rules Group

The _"id_tax_rules"_ is requested by the function **_Product::getIdTaxRulesGroupByIdProduct()_** and used cache to avoid multiples queries.
This data is already present in the global query.

####  Price

Price computing is a big thing, the internal Prestashop methods use 2 levels of cache, one for attributes prices (_$_pricesLevel2_) before computing and one (_$_prices_) for the final price.
The global query results contains almost all data to constitute the default combination price cache, only _"attribute_price"_, _"default_on"_, _"attribute_ecotax"_ are missing.

####  Quantities

Quantity computing is a big thing too, the process call **_Product::getQuantity()_** which call **_StockAvailable::getQuantityAvailableByProduct()_**.
The global query contains all data to constitute the default combination quantity cache. 
Notable fact or bug, the cache_id is set with a null shop in this context.

#### Category Name

The _"category_name"_ is loaded by a "raw" SQL query for each product.
```php
$row['category_name'] = Db::getInstance()->getValue('SELECT name FROM ....);
```

A simple cache can be used to avoid the same request multiples times.

#### GroupReduction

To determine product price, the system check if the current user have a group reduction for this product, a way of improvement is to load all the reduction for a given group (with id_product filters), put on a new "inverted" cache, and when **_GroupReduction::getValueForProduct()_** is called check in this inverted cache if group_reduction was already loaded and if a reduction is present for this product.

#### Cart & Order

To determine product quantity, the system check if the cart is empty and if there is an order with this id_cart.
First, It's possible to cache the **_Order::getByCartId()_** result.
Second, it's possible to ignore some queries when the cart is empty.

#### Attributes ColorList

In my cases, the attributes color list is never used for my products, but the system try to load it for each products.

#### Attributes features

I used attributes features in the product page or the quick view, but the category listing never display it, so this loading is unnecessary.

#### Product Assembler

The class **_ProductAssembler_** is used to prepare the product data just before to be rendered. The assembling process is composed by a big query (**_addMissingProductsFields()_**) and a re-call of **_Product::getProperties()_**.

I analyze the difference between the category global query and the result and only missed 3 data to avoid this big query (_"id_lang"_, _"delivery_in_stock"_, _"delivery_out_stock"_).
It's possible to bypass this step if we load the 3 missed data in the global category query.

#### ImageRetriever & Link

This objects can re-instanciate **Product** object and load images & combinations images. In my cases, i don't use combinations but a lot of empty queries are made.

## Proof of concept

I wrote a simple override for Product.php which try to pre-cache the products rows before call Product::getProductsProperties(). My category contains 50 products.

### Code 
```php
class Product extends ProductCore {

    public const MODE_PRECACHING = true;

    public static function getProductsProperties($id_lang, $query_result)
    {
        if(self::MODE_PRECACHING) {
            foreach ($query_result as &$row) {
                if (array_key_exists('id_image', $row)) {
                    // precaching cover
                    $cache_id = 'Product::getCover_' . (int)$row['id_product'] . '-' . (int)$row['id_shop'];
                    $precacheValue = [$row['id_image']];
                    if (!Cache::isStored($cache_id)) {
                        Cache::store($cache_id, $precacheValue);
                    }
                }

                if(array_key_exists('id_tax_rules_group', $row)) {
                    // precaching tax rules group
                    $cache_id = 'product_id_tax_rules_group_' . (int) $row['id_product'] . '_' . $row['id_shop'];
                    $precacheValue = $row['id_tax_rules_group'];
                    if (!Cache::isStored($cache_id)) {
                        Cache::store($cache_id, $precacheValue);
                    }
                }

                if(array_key_exists('price', $row)) {
                    // precaching attributes prices level2,
                    $cacheLevel2Id = $row['id_product'] . '-' . $row['id_shop'];
                    $precacheValue = [
                        'price' => $row['price'],
                        'ecotax' => $row['ecotax'],
                        'id_product_attribute' => $row['id_product_attribute'],
                        'attribute_price' => null, // missed on global request, set null by default
                        'default_on' => null, // missed on global request, set null by default
                        'attribute_ecotax' => null, // missed on global request, set null by default
                    ];
                    self::$_pricesLevel2[$cacheLevel2Id][(int) $row['id_product_attribute']] = $precacheValue;
                }

                if(array_key_exists('quantity', $row)) {
                    // precaching quantity
                    $prefix_cache_id = 'StockAvailable::getQuantityAvailableByProduct_' . (int)$row['id_product'] . '-' . (int)$row['id_product_attribute'] . '-';
                    $cache_id = $prefix_cache_id . (int)$row['id_shop'];
                    $precacheValue = (int)$row['quantity'];
                    if (!Cache::isStored($cache_id)) {
                        Cache::store($cache_id, $precacheValue);
                    }
                    // Probable bug, cache id compute by the classical way is for shop 0, fix this value
                    $fixed_cached_id = $prefix_cache_id . '0';
                    if (!Cache::isStored($fixed_cached_id)) {
                        Cache::store($fixed_cached_id, $precacheValue);
                    }
                }
            }
        }
        return parent::getProductsProperties($id_lang, $query_result);
    }
}
```

### Without optimization

![Capture d’écran du 2023-06-30 11-55-01](https://github.com/PrestaShop/PrestaShop/assets/16869488/941f1a28-64dd-4ce6-be21-39fe287184b1)

### With Pre-caching only

![Capture d’écran du 2023-06-30 11-55-38](https://github.com/PrestaShop/PrestaShop/assets/16869488/94e62257-99d9-4b94-9265-37274907d500)

### With all optimizations

To do this, i develop specific overrides on my store to resolve all listed optimization ideas : GroupReduction, Order, Cart, ProductAssembler, Link and i used an extended class of ImageRetriever to used only cover already image and ignore combination images.

![Capture d’écran du 2023-06-30 11-56-49](https://github.com/PrestaShop/PrestaShop/assets/16869488/d45256cc-f9eb-45d6-9b44-f5067f6497ae)

With all optimization, the metrics are very good but some development is very specific to my config store (no combinations, no features, no attributes colors...).

### Do you plan to work on this subject?

- [ ] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [ ] I'm willing to provide any wireframes or design assets required for this feature.
- [ ] I'm willing to submit a Pull Request that implements this feature.
- [X] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

**Comment by Hlavtox** [comment]
Hi @MaximeOh, I work on this gradually.

For 8 or 8.1, can't remember now:
- I optimized ProductAssembler to load data in one query, as a first step.
- Then I did some caching and speedup in faceted search by caching features and attribtes there.

For develop currently:
- Now, features are lazyloaded, saves another 24 queries, https://github.com/PrestaShop/PrestaShop/pull/33335
- Attribute price lazyloaded if needed https://github.com/PrestaShop/PrestaShop/pull/33435
- Removed unnecessary loading of image covers here - https://github.com/PrestaShop/PrestaShop/pull/33363
- I will continue with category and manufacturer info. 👍

I also tried to experiment with lazyloading data for the whole list in one query. https://github.com/PrestaShop/PrestaShop/pull/33381
So the code would know that if there is one hit, there will hits probably for the whole list, and it will fetch it for the whole list at once.

**Comment by Hlavtox** [comment]
@MaximeOh More things implemented here - https://github.com/PrestaShop/PrestaShop/pull/34499

After it's merged, I will continue with other things. :-)
